### PR TITLE
Add compiler support for UnconditionalSuppressMessage

### LINKED ISSUE #48885: Request: Treat `UnconditionalSuppressMessageAttribute` like `SuppressMessageAttribute`
History: `SuppressMessageAttribute` has a `ConditionalAttribute` on it so it doesn't get emitted. But tools like the linker need a way for code to suppress warnings and a way to see those suppressions in the final binaries.

[UnconditionalSuppressMessage](https://source.dot.net/#System.Private.CoreLib/UnconditionalSuppressMessageAttribute.cs,a81b3979c947c241,references) was added for this purpose.

We're now trying to add analyzers to produce the same warnings that the linker produces, but at build time. If the analyzer infrastructure doesn't recognize `UnconditionalSuppress` users will have to annotate their code twice (and it will cause new previously suppressed warnings to appear when we turn on the analyzer).

Overall, we think this is critical for the linker to be successful in the .NET 6 epic of producing a good trimming experience (https://github.com/dotnet/runtime/issues/43543).

### PR DESCRIPTION
**Pull request by agocke**
Fixes #48885

### CODE REVIEW

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/SuppressMessageAttributeCompilerTests.cs
Comment (AlekseyTs) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/SuppressMessageAttributeCompilerTests.cs:62]: >mscorlibRef [](start = 28, length = 11)

Taking dependency on particular version of core library feels fragile.

Comment (AlekseyTs) [reply src/Compilers/Core/CodeAnalysisTest/Diagnostics/SuppressMessageAttributeCompilerTests.cs:62]: I see that we are already doing this below. Consider using this local in all places to make sure consistency across all compilations.

---
In reply to: [595318758](https://github.com/dotnet/roslyn/pull/51792#discussion_r595318758) [](ancestors = 595318758)

Comment (AlekseyTs) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/SuppressMessageAttributeCompilerTests.cs:62]: >EmitToImageReference [](start = 41, length = 20)

Consider caching this reference and reusing across multiple calls of ```CreateCompilation```.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: >GetTypeByMetadataName [](start = 78, length = 21)

I think this API can return null, so using null as a sentinel for ```_lazyUnconditionalSuppressMessageAttribute``` is probably not optimal.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:267]: It looks like this condition is used in at least two places, consider extracting and reusing a helper.

File: src/EditorFeatures/Test/Diagnostics/SuppressMessageAttributeWorkspaceTests.cs
Comment (AlekseyTs) [inline src/EditorFeatures/Test/Diagnostics/SuppressMessageAttributeWorkspaceTests.cs:53]: >TestBase.MscorlibRef [](start = 38, length = 20)

How do we know this is the right version to use?

File: src/EditorFeatures/Test/Diagnostics/SuppressMessageAttributeWorkspaceTests.cs
Comment (AlekseyTs) [inline src/EditorFeatures/Test/Diagnostics/SuppressMessageAttributeWorkspaceTests.cs:58]: >EmitToImageReference [](start = 41, length = 20)

Is it possible to cache/reuse this result?

File: src/EditorFeatures/Test/Diagnostics/SuppressMessageAttributeWorkspaceTests.cs
Comment (AlekseyTs) [inline src/EditorFeatures/Test/Diagnostics/SuppressMessageAttributeWorkspaceTests.cs:65]: >refComp [](start = 70, length = 7)

Are there actually tests that take advantage of this reference?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: This should be implemented according to the design in https://github.com/dotnet/roslyn/pull/51571

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: > This should be implemented according to the design in [https://github.com/dotnet/roslyn/pull/51571](https://github.com/dotnet/roslyn/pull/51571)

Do you think it isn't?


---
In reply to: [595559328](https://github.com/dotnet/roslyn/pull/51792#discussion_r595559328) [](ancestors = 595559328)

Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: It seems to only apply to `GetWellKnownType`?

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: > It seems to only apply to GetWellKnownType?

Are you judging by PR description? I think it is imprecise, to narrow. Check out the tests in the PR.

---
In reply to: [595589343](https://github.com/dotnet/roslyn/pull/51792#discussion_r595589343) [](ancestors = 595589343)

Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: In this case, I'd be looking for a test like the following:

* Assembly TA declares the attribute as `internal`
* Assembly TB declared the attribute as `public`
* Assembly TC references TA and TB, and uses the attribute

Verify that the suppression, which resolves to the attribute declared in TB, works properly.

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: > Verify that the suppression, which resolves to the attribute declared in TB, works properly.

I don't think GetTypeByMetadataName cares about accessibility. In any case, this component already relies on GetTypeByMetadataName API for System.Diagnostics.CodeAnalysis.SuppressMessageAttribute. If you think lookup should be performed differently, I would say both use-sites should be adjusted and that is probably out of scope for this PR.

---
In reply to: [595600646](https://github.com/dotnet/roslyn/pull/51792#discussion_r595600646) [](ancestors = 595600646)

Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: SuppressMessageAttribute was added in .NET 2.0, so there wouldn't be as much of a need for someone to declare their own copy. The new attribute is much more likely to be impacted.

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: > The new attribute is much more likely to be impacted.

Perhaps. 
1) Rather than trying to infer if the behavior matches your expectations from tests, it would probably be better if you could specify the behavior you want first.
2) I think we might still want to use the same lookup rules for both attributes. 
3) Andy is probably not the right person to implement new lookup rules (if we really want new lookup rules). Should someone from Analyzers team take over this PR? Or can this be done as a follow-up?

---
In reply to: [595615767](https://github.com/dotnet/roslyn/pull/51792#discussion_r595615767) [](ancestors = 595615767)

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: For example, you might want to switch to the way compilers recognize attributes - fully qualified name of the type + signature of the constructor used. The lookup of a "blessed" type is not involved and that avoids the need to deal with all kinds of possible lookup ambiguities. An attribute with the right name and signature is applied, that is the "right" attribute and it should be respected. 

---
In reply to: [595619781](https://github.com/dotnet/roslyn/pull/51792#discussion_r595619781) [](ancestors = 595619781,595615767)

File: src/EditorFeatures/Test/Diagnostics/SuppressMessageAttributeWorkspaceTests.cs
Comment (agocke) [reply src/EditorFeatures/Test/Diagnostics/SuppressMessageAttributeWorkspaceTests.cs:53]: I think this should be OK regardless: all 4.x+ corlibs should be unified, and I'm only using AttributeUsage, Attribute, and string. Those should always be present.

I looked at trying to find the references used by the workspaces layer, but the TestWorkspace was confusing and I couldn't figure out a good way to keep them in-sync.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (agocke) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs]: Yeah, @AlekseyTs has a good point here -- I was actually surprised that this code looks for a particular type definition for attributes. In the compiler attributes are often accepted as long as they have the right name and signature, regardless of where they are defined.

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/SuppressMessageAttributeCompilerTests.cs
Comment (AlekseyTs) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/SuppressMessageAttributeCompilerTests.cs:35]: >new [](start = 87, length = 3)

Please avoid using target typed `new` under Compilers.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:226]: >new [](start = 52, length = 3)

Please don't use target typed `new`.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: >new [](start = 65, length = 3)

Please don't use target typed `new`.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: >_lazyUnconditionalSuppressMessageAttribute [](start = 20, length = 42)

Is this assignment thread-safe?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:226]: >_lazySuppressMessageAttribute [](start = 20, length = 29)

Is this assignment thread-safe?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (agocke) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:239]: I think so, in the specific way I was using it -- but rather than guess I just changed it to something I'm certain is safe.

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs]: > I think so, in the specific way I was using it

It looks like Optional is a struct with several fields, the assignment is not guaranteed to be atomic. What am I missing?

---
In reply to: [597090036](https://github.com/dotnet/roslyn/pull/51792#discussion_r597090036) [](ancestors = 597090036)

Comment (agocke) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs]: You're right, I the previous version wasn't thread safe. The path would be:

Thread 1:
_field.HasValue is false
retrieve Type
_field.HasValue <- true
--- interrupted here

Thread 2 (continuing at interruption):
_field.HasValue is true
return value is null (not correct)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5), assuming CI is passing

**Comment by agocke** [comment]
Looks like there was an infra break -- I don't think these were caused by my changes.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:240]: Just checking my understanding: is the purpose of this StrongBox just to allow us to distinguish these cases:
1. _lazySuppressMessageAttribute is not initialized
2. _lazySuppressMessageAttribute is initialized, and its value is null
3. _lazySuppressMessageAttribute is initialized, and its value is non-null

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:312]: Could this be called with an argument with a null `AttributeClass`? I'm wondering if this API could give strange answers in the event that either SuppressMessageAttribute or UnconditionalSuppressMessageAttribute is null, and `a.AttributeClass` is null.

**Review by RikkiGibson — COMMENTED** [review]
Looks good, just had a few small questions.

Is @dotnet/roslyn-ide sign-off needed for the EditorFeatures test changes?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (agocke) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:240]: Correct, although we don't really need to care whether the value is null or non null, just whether or not it's initialized.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (agocke) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:312]: Heh, yeah I suppose that's possible. I think that's an existing bug. I'd rather not hold up the PR for it if you don't mind, but I can file a bug about it.

Comment (agocke) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:312]: (I don't think it semantically matters because the diagnostic is only suppressed if the ID's match, but if the attribute doesn't exist, we won't be able to fetch the ID and they will never match).

Comment (agocke) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:312]: Filed https://github.com/dotnet/roslyn/issues/51972

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
Comment (RikkiGibson) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs:312]: Sounds good.

**Comment by agocke** [comment]
@sharwell Anything else? I'd like to get this in for Preview 3 (snaps tomorrow)

**Comment by sharwell** [comment]
Send me a message tomorrow on teams so I don't miss it

**Review by sharwell — APPROVED** [review]
Looks fine with the understanding that this will break for cases where a dependency declares a second copy of the attribute as `internal` (even if there is no IVT access to it).

**Comment by RikkiGibson** [comment]
Are we good to merge this? @agocke do you have a preference for squash merge or merge commit?

**Comment by agocke** [comment]
Squashed, thanks.

**Pull request merged**
